### PR TITLE
Modify some oneflow-documentation errors

### PR DESCRIPTION
**Pull request by process852**
文档校正后的[在线浏览链接](https://process852.github.io/oneflow-documentation/)

* 修正由于[API](https://github.com/Oneflow-Inc/oneflow/issues/9457)改动造成的[文档](https://docs.oneflow.org/master/)错误
    - https://github.com/Oneflow-Inc/oneflow/issues/9457
* 改动后代码块的测试结果https://github.com/Oneflow-Inc/oneflow-documentation/pull/517#issuecomment-1345828801
* 改动界面主要是以下几个
1. [流水并行训练](https://process852.github.io/oneflow-documentation/master/parallelism/06_pipeline.html)
2. [使用 Global Tensor 进行分布式编程：分布式并行策略](https://process852.github.io/oneflow-documentation/master/cookies/global_tensor_distributed.html)
3. [Activation Checkpointing](https://process852.github.io/oneflow-documentation/master/cookies/activation_checkpointing.html)
4. [大模型分片保存和加载](https://process852.github.io/oneflow-documentation/master/cookies/save_load.html)

### CODE REVIEW

**Comment by doombeaker** [comment]
麻烦贴一个在线版，方便 reviewer 预览吧

**Comment by process852** [comment]
* 针对原先提交补充一些问题改正
* 原先提交内容回顾浏览: https://github.com/Oneflow-Inc/oneflow-documentation/pull/517#issue-1483501796
* 文档校正后的[在线浏览链接](https://process852.github.io/oneflow-documentation/)
* 新提交的commit针对`requirements.txt`进行补充
    - [Pygments](https://pygments.org/docs/changelog/)是一个将代码格式化的Python库,可以将代码格式化为html代码、图片、rtf文件等多种格式
    - 当使用低版本的文件包时会出现网页端代码片段显示不正常的情况，情况如下图所示
image
    - 正常显示的情况应该如下:
image
    - 故建议使用`Pygments >= 2.13.0`及以上版本，`requirements.txt`使用的是`Pygments == 2.13.0`

**Review by lixiang007666 — APPROVED** [review]

**Comment by process852** [comment]
* 在执行自动化脚本`run_markdown_file`检查时出现代码运行错误[错误链接](https://github.com/Oneflow-Inc/oneflow-documentation/actions/runs/3672092690/jobs/6207946360)
image
* 现定位到错误发生在[Embedding章节](https://docs.oneflow.org/master/cookies/one_embedding.html)
* 文档中原始代码如下:
```python
import oneflow as flow
import oneflow.nn as nn
import numpy as np

tables = [
    flow.one_embedding.make_table_options(
        flow.one_embedding.make_uniform_initializer(low=-0.1, high=0.1)
    ),
    flow.one_embedding.make_table_options(
        flow.one_embedding.make_uniform_initializer(low=-0.05, high=0.05)
    ),
    flow.one_embedding.make_table_options(
        flow.one_embedding.make_uniform_initializer(low=-0.15, high=0.15)
    ),
]

store_options = flow.one_embedding.make_cached_ssd_store_options(
    cache_budget_mb=8142,
    persistent_path="./your_path_to_ssd", 
    capacity=40000000,
    size_factor=1,              
    physical_block_size=512
)

embedding_size = 128
embedding = flow.one_embedding.MultiTableEmbedding(
    name="my_embedding",
    embedding_dim=embedding_size,
    dtype=flow.float,
    key_type=flow.int64,
    tables=tables,
    store_options=store_options,
)

embedding.to("cuda")

num_tables = 3
mlp = flow.nn.FusedMLP(
    in_features=embedding_size * num_tables,
    hidden_features=[512, 256, 128],
    out_features=1,
    skip_final_activation=True,
)
mlp.to("cuda")

class TrainGraph(flow.nn.Graph):
    def __init__(self,):
        super().__init__()
        self.embedding_lookup = embedding
        self.mlp = mlp
        self.add_optimizer(
            flow.optim.SGD(self.embedding_lookup.parameters(), lr=0.1, momentum=0.0)
        )
        self.add_optimizer(
            flow.optim.SGD(self.mlp.parameters(), lr=0.1, momentum=0.0)
        )
    def build(self, ids):
        embedding = self.embedding_lookup(ids)
        loss = self.mlp(flow.reshape(embedding, (-1, num_tables * embedding_size)))
        loss = loss.sum()
        loss.backward()
        return loss


ids = np.random.randint(0, 1000, (100, num_tables), dtype=np.int64)
ids_tensor = flow.tensor(ids, requires_grad=False).to("cuda")
graph = TrainGraph()
loss = graph(ids_tensor)
print(loss)
```
* 调试过程中发现在`build`函数中执行`loss.backward()`时会出现如下图所示错误
image
* 版本信息
```
oneflow                   0.8.1.dev20221201+cu112        
numpy                     1.23.4                  
```

**Comment by process852** [comment]
* 本地自动化CI测试脚本，仓库原生配置的`yml`文件都没有涉及到分布式启动的文档内容
* 现在希望可以添加代码实现分布式训练代码块的CI测试
* 基本想法:将代码块保存为一个临时文件，然后使用`os.system`启动分布式训练
```python
with open('temp.py', 'w') as f:
    # print(block_dict[index])
    f.write(block_dict[index])
    temppath = os.path.join(os.getcwd(), 'temp.py')
    os.system('python3 -m oneflow.distributed.launch --nproc_per_node 2 {}'.format(temppath))
    os.remove(temppath)
```
* 测试结果：没有报错,但是`os.system`无法显示执行脚本的输出信息
image

**Comment by process852** [comment]
* 修改的`markdown`文件本地代码块CI测试结果
* [Activation Checkpointing](https://process852.github.io/oneflow-documentation/master/cookies/activation_checkpointing.html) 
   测试命令：`python  run_by_yamls.py --yaml ./configs/guide_activation_checkpointing.yml`
测试结果：
image

* 以下几个代码块修改涉及分布式启动，还未配置相关测试代码，故在本地直接运行给出结果
* [使用 Global Tensor 进行分布式编程：分布式并行策略](https://process852.github.io/oneflow-documentation/master/cookies/global_tensor_distributed.html) 
代码块：
```python
import oneflow as flow
import oneflow.nn as nn

P01 = flow.placement(type="cuda", ranks=[0, 1])
P23 = flow.placement(type="cuda", ranks=[2, 3])


class StageModule(nn.Module):
    def __init__(self, in_dims, out_dims, placement=None, sbp=None):
        super().__init__()
        self.w = nn.Parameter(
            flow.randn(in_dims, out_dims, placement=placement, sbp=sbp)
        )

    def forward(self, x):
        out = flow.matmul(x, self.w)
        return out


class ModuleModel(nn.Module):
    def __init__(self):
        super().__init__()

        # 模型第一阶段在第 0 和第 1 卡上进行数据并行计算
        self.m_stage0 = StageModule(5, 8, placement=P01, sbp=flow.sbp.broadcast)

        # 模型第二阶段在第 2 和第 3 卡上进行模型并行计算
        self.m_stage1 = StageModule(8, 3, placement=P23, sbp=flow.sbp.split(dim=1))

    def forward(self, x):
        # 第一阶段，数据切分在第 0 和第 1 卡，用于数据并行
        out_stage0 = self.m_stage0(x)

        # 第二阶段需要将输入数据还原完整，并转移至第 2 和第 3 卡，用于模型并行
        in_stage1 = out_stage0.to_global(placement=P23, sbp=flow.sbp.broadcast)
        out_stage1 = self.m_stage1(in_stage1)

        return out_stage0, out_stage1


# Graph
class GraphModel(nn.Graph):
    def __init__(self):
        super().__init__()
        self.model = ModuleModel()
        self.model.m_stage0.to(nn.graph.GraphModule).set_stage(stage_id=0, placement=P01)
        self.model.m_stage1.to(nn.graph.GraphModule).set_stage(stage_id=1, placement=P23)

    def build(self, x):
        return self.model(x)


if __name__ == "__main__":
    graph = GraphModel()
    # 需要将输入数据切分，用于数据并行
    in_stage0 = flow.randn(4, 5, placement=P01, sbp=flow.sbp.split(dim=0))
    out_stage0, out_stage1 = graph(in_stage0)
    print(out_stage0.shape, out_stage1.shape)  # (4, 8) (4, 3)
```
测试命令：`python3 -m oneflow.distributed.launch --nproc_per_node 4 mixparallel.py`
测试结果:
image
* [大模型分片保存和加载](https://process852.github.io/oneflow-documentation/master/cookies/save_load.html)
测试命令:`python3 -m oneflow.distributed.launch --nproc_per_node 4 Bigmodule.py` 
测试结果：
image
image
* [流水并行训练](https://process852.github.io/oneflow-documentation/master/parallelism/06_pipeline.html)
测试命令: `python3 -m oneflow.distributed.launch --nproc_per_node 2  pipeline.py `
测试结果：
image

File: cn/docs/cookies/activation_checkpointing.md [hunk 1]
```diff
@@ -40,7 +40,7 @@ optimizer = flow.optim.SGD([{'params': model_part1.parameters()},
                            lr=1e-3)
 ```
 
-如果要开启 activation checkpointing，只需在 [nn.Graph](../basics/08_nn_graph.md) 模型中的 Eager 模型成员 (即 nn.Module 对象) 上指定 `.config.activation_checkpointing = True`。此 API 详见：[activation_checkpointing](https://oneflow.readthedocs.io/en/v0.8.1/generated/oneflow.nn.graph.block_config.BlockConfig.activation_checkpointing.html)。对于每个打开 "activation checkpointing" 的 nn.Module，其输入 activation 将会被保留，而其它中间 activation 在反向传播过程中被使用时会被重新计算。
+如果要开启 activation checkpointing，只需在 [nn.Graph](../basics/08_nn_graph.md) 模型中的 Eager 模型成员 (即 nn.Module 对象) 利用`.to(nn.graph.GraphModule)`方法转换为`nn.graph.GraphModule`对象，并在其上指定 `.activation_checkpointing = True`。此 API 详见：[activation_checkpointing](https://oneflow.readthedocs.io/en/v0.8.1/generated/oneflow.nn.graph.block_config.BlockConfig.activation_checkpointing.html)。对于每个打开 "activation checkpointing" 的 nn.Module，其输入 activation 将会被保留，而其它中间 activation 在反向传播过程中被使用时会被重新计算。
```
Comment (doombeaker) [inline cn/docs/cookies/activation_checkpointing.md:43]:
```suggestion
如果要开启 activation checkpointing，只需在 [nn.Graph](../basics/08_nn_graph.md) 模型中的 Eager 模型成员 (即 nn.Module 对象) 利用 `.to(nn.graph.GraphModule)` 方法转换为 `nn.graph.GraphModule` 对象，并在其上指定 `.activation_checkpointing = True`。此 API 详见：[activation_checkpointing](https://oneflow.readthedocs.io/en/v0.8.1/generated/oneflow.nn.graph.block_config.BlockConfig.activation_checkpointing.html)。对于每个打开 "activation checkpointing" 的 `nn.Module`，其输入 activation 将会被保留，而其它中间 activation 在反向传播过程中被使用时会被重新计算。
```

中英文之间需要空格。这个是我们的文档里尽量遵循的规范（没遵循的是因为其它原因，降低了标准）

File: cn/docs/parallelism/06_pipeline.md [hunk 2]
```diff
@@ -144,12 +144,13 @@ P1 = flow.placement("cuda", ranks=[1])
 
 ### Stage ID 及梯度累积设置
 
-通过 Module Config 上的 [config.set_stage](https://oneflow.readthedocs.io/en/v0.8.1/generated/oneflow.nn.graph.block_config.BlockConfig.set_stage.html) 方法，设置流水线 Stage ID 和 Stage 对应的 Placement，Stage ID 从 0 开始编号，依次加 1。
+当[nn.Module](https://oneflow.readthedocs.io/en/master/nn.html?highlight=nn.Module#nn-module)的一个实例化网络层作为属性加入继承于[nn.Graph](https://oneflow.readthedocs.io/en/master/graph.html)的新类时，内部会将该网络层用`ProxyModule`进行包装，利用方法`.to`得到一个`nn.graph.GraphModule`的实例化对象，然后使用方法`stage_id`设置流水线 Stage ID 和 Stage 对应的 Placement，Stage ID 从 0 开始编号，依次加 1。
```
Comment (doombeaker) [inline cn/docs/parallelism/06_pipeline.md:147]: 同上，中英文间的空格检查、补一下吧。

**Review by doombeaker — APPROVED** [review]

File: cn/docs/cookies/activation_checkpointing.md (hunk 1 above)
Comment (process852) [reply cn/docs/cookies/activation_checkpointing.md]: 已检查

**Review by strint — APPROVED** [review]
